File: backend/src/mhglauncher/services/uigf.py
from __future__ import annotations

from datetime import UTC, datetime
from typing import Any, Literal

from pydantic import BaseModel, ConfigDict, Field, ValidationError, field_validator

from mhglauncher import __version__
from mhglauncher.errors import AppError
from mhglauncher.models import WishRecord
from mhglauncher.services.item_metadata import enrich_record

UIGFVersion = Literal["v4.0", "v4.1", "v4.2"]
GACHA_TYPES = {"100", "200", "301", "302", "400", "500"}
UIGF_GACHA_TYPES = {"100", "200", "301", "302", "500"}


class UIGFInfo(BaseModel):
    model_config = ConfigDict(extra="ignore")

    export_timestamp: str | int
    export_app: str
    export_app_version: str
    version: UIGFVersion


class UIGFItem(BaseModel):
    model_config = ConfigDict(extra="ignore")

    uigf_gacha_type: str
    gacha_type: str
    item_id: str
    count: str = "1"
    time: str
    name: str = ""
    item_type: str = ""
    rank_type: str | None = None
    id: str = Field(min_length=1, max_length=19, pattern=r"^[0-9]+$")

    @field_validator(
        "uigf_gacha_type",
        "gacha_type",
        "item_id",
        "count",
        "rank_type",
        "id",
        mode="before",
    )
    @classmethod
    def stringify_numbers(cls, value: Any) -> Any:
        return str(value) if isinstance(value, int) else value

    @field_validator("uigf_gacha_type")
    @classmethod
    def validate_uigf_gacha_type(cls, value: str) -> str:
        if value not in UIGF_GACHA_TYPES:
            raise ValueError("unsupported UIGF gacha type")
        return value

    @field_validator("gacha_type")
    @classmethod
    def validate_gacha_type(cls, value: str) -> str:
        if value not in GACHA_TYPES:
            raise ValueError("unsupported gacha type")
        return value


class UIGFAccount(BaseModel):
    model_config = ConfigDict(extra="ignore")

    uid: str
    timezone: int
    lang: str | None = None
    list: list[UIGFItem]

    @field_validator("uid", mode="before")
    @classmethod
    def stringify_uid(cls, value: Any) -> Any:
        return str(value) if isinstance(value, int) else value


class UIGFFile(BaseModel):
    model_config = ConfigDict(extra="ignore")

    info: UIGFInfo
    hk4e: list[UIGFAccount] = Field(default_factory=list)


def import_uigf(payload: dict[str, Any]) -> list[WishRecord]:
    _reject_legacy_version(payload)
    try:
        document = UIGFFile.model_validate(payload)
        records = [_record(account, item) for account in document.hk4e for item in account.list]
    except ValidationError as error:
        raise AppError("uigf_invalid", "UIGF 文件不符合 v4.0、v4.1 或 v4.2 规范") from error
    if not records:
        raise AppError("uigf_empty", "UIGF 文件不包含原神祈愿记录")
    return records


def export_uigf(uid: str, records: list[WishRecord]) -> dict[str, Any]:
    return {
        "info": {
            "export_timestamp": int(datetime.now(UTC).timestamp()),
            "export_app": "MHGLauncher",
            "export_app_version": __version__,
            "version": "v4.2",
        },
        "hk4e": [
            {
                "uid": uid,
                "timezone": _timezone(uid),
                "lang": "zh-cn",
                "list": [_export_item(item) for item in reversed(records)],
            }
        ],
    }


def _record(account: UIGFAccount, item: UIGFItem) -> WishRecord:
    try:
        parsed_time = datetime.strptime(item.time, "%Y-%m-%d %H:%M:%S")
        rank = int(item.rank_type) if item.rank_type is not None else 0
        return enrich_record(
            WishRecord(
                id=item.id,
                uid=account.uid,
                gacha_type=item.gacha_type,
                uigf_gacha_type=item.uigf_gacha_type,
                item_id=item.item_id,
                name=item.name,
                item_type=item.item_type,
                rank=rank,
                time=parsed_time,
            )
        )
    except (TypeError, ValueError) as error:
        raise AppError("uigf_item_invalid", "UIGF 记录字段无效") from error


def _export_item(item: WishRecord) -> dict[str, str]:
    result = {
        "uigf_gacha_type": item.uigf_gacha_type or _uigf_type(item.gacha_type),
        "gacha_type": item.gacha_type,
        "item_id": item.item_id,
        "count": "1",
        "time": item.time.strftime("%Y-%m-%d %H:%M:%S"),
        "id": item.id,
    }
    if item.name:
        result["name"] = item.name
    if item.item_type:
        result["item_type"] = item.item_type
    if item.rank:
        result["rank_type"] = str(item.rank)
    return result


def _reject_legacy_version(payload: dict[str, Any]) -> None:
    info = payload.get("info")
    if isinstance(info, dict) and "uigf_version" in info and "version" not in info:
        raise AppError("uigf_legacy", "仅支持 UIGF v4.0、v4.1、v4.2，请先升级旧版文件")


def _uigf_type(gacha_type: str) -> str:
    return "301" if gacha_type == "400" else gacha_type


def _timezone(uid: str) -> int:
    if uid.startswith("6"):
        return -5
    if uid.startswith("7"):
        return 1
    return 8

File: backend/src/services/uigf.ts
import { z } from "zod";
import { AppError } from "../core/errors";
import type { WishRecord } from "../core/models";
import { enrich } from "./metadata";

const item = z.object({
  id: z.coerce.string().regex(/^\d{1,19}$/), uigf_gacha_type: z.coerce.string(),
  gacha_type: z.coerce.string(), item_id: z.coerce.string(), time: z.string(),
  name: z.string().default(""), item_type: z.string().default(""),
  rank_type: z.coerce.string().optional(),
}).passthrough();
const gameUid = z.coerce.string().regex(/^\d{9,10}$/);
const account = z.object({ uid: gameUid, timezone: z.coerce.number().int().min(-12).max(14).default(8), list: z.array(item) }).passthrough();
const modern = z.object({ info: z.object({ version: z.enum(["v4.0", "v4.1", "v4.2"]) }).passthrough(), hk4e: z.array(account).default([]) }).passthrough();
const legacy = z.object({
  info: z.object({ uid: gameUid, uigf_version: z.string().regex(/^v[23]\./) }).passthrough(),
  list: z.array(item).default([]),
}).passthrough();
const gachaTypes = new Set(["100", "200", "301", "302", "400", "500"]);
const uigfTypes = new Set(["100", "200", "301", "302", "500"]);

export function importUIGF(payload: unknown): WishRecord[] {
  try {
    const raw = payload as { info?: { uigf_version?: string } };
    const groups: { uid: string; timezone: number; list: z.infer<typeof item>[] }[] = raw.info?.uigf_version
      ? ((value) => [{ uid: value.info.uid, timezone: 8, list: value.list }])(legacy.parse(payload))
      : modern.parse(payload).hk4e;
    const records = groups.flatMap((group) => group.list.map((value) => record(group.uid, group.timezone, value)));
    if (!records.length) throw new AppError("uigf_empty", "UIGF 文件不包含原神祈愿记录");
    return records;
  } catch (error) {
    if (error instanceof AppError) throw error;
    throw new AppError("uigf_invalid", "UIGF 文件不符合受支持的规范");
  }
}

export function exportUIGF(uid: string, records: WishRecord[]): Record<string, unknown> {
  const timezone = uid.startsWith("6") ? -5 : uid.startsWith("7") ? 1 : 8;
  return {
    info: { export_timestamp: Math.floor(Date.now() / 1000), export_app: "MHGLauncher", export_app_version: "1.0.0", version: "v4.2" },
    hk4e: [{ uid, timezone, lang: "zh-cn", list: records.toReversed().map((value) => exportItem(value, timezone)) }],
  };
}

function record(uid: string, timezone: number, value: z.infer<typeof item>): WishRecord {
  const time = normalizeTime(value.time, timezone);
  if (!gachaTypes.has(value.gacha_type) || !uigfTypes.has(value.uigf_gacha_type) || !time) {
    throw new AppError("uigf_item_invalid", "UIGF 记录字段无效");
  }
  return enrich({
    id: value.id, uid, gacha_type: value.gacha_type, uigf_gacha_type: value.uigf_gacha_type,
    item_id: value.item_id, name: value.name, item_type: value.item_type,
    rank: Number(value.rank_type ?? 0), time,
  });
}

function normalizeTime(value: string, timezone: number): string | null {
  const candidate = value.trim().replace(" ", "T");
  const offset = timezone >= 0 ? `+${String(timezone).padStart(2, "0")}:00` : `-${String(-timezone).padStart(2, "0")}:00`;
  const source = /(?:Z|[+-]\d{2}:?\d{2})$/i.test(candidate) ? candidate : `${candidate}${offset}`;
  const epoch = Date.parse(source);
  return Number.isFinite(epoch) ? new Date(epoch).toISOString() : null;
}

function exportItem(value: WishRecord, timezone: number): Record<string, string> {
  const epoch = Date.parse(value.time);
  const local = new Date(epoch + timezone * 3_600_000).toISOString().replace("T", " ").slice(0, 19);
  const result: Record<string, string> = {
    uigf_gacha_type: value.uigf_gacha_type || (value.gacha_type === "400" ? "301" : value.gacha_type),
    gacha_type: value.gacha_type, item_id: value.item_id, count: "1", time: local, id: value.id,
  };
  if (value.name) result.name = value.name;
  if (value.item_type) result.item_type = value.item_type;
  if (value.rank) result.rank_type = String(value.rank);
  return result;
}
